### DISCUSSION TITLE
Test coverage workflows

### DISCUSSION THREAD

**Discussion by keith** [MobileNativeFoundation/discussions#38] [General]
Not to start another thread on testing 🙃 but I'm curious what tools / processes folks use around code coverage.

At Lyft we have a few things:

- We let developers set a minimum required coverage % on their code if they want to, and we fail CI if coverage drops below that
- We provide some local workflow to view coverage which can be helpful for backfilling coverage in places where it's missing
- We provide a web UI that can show coverage from PRs or overall, but it's not well integrated in to GitHub or anything

We don't really use coverage as a top level metric for much, just another indicator for quality, but it's mostly a developer facing tool to understand areas that aren't tested but potentially could be.

**Comment by dflems** [comment]
Same with the minimum coverage enforcement. We also have a custom web UI that we built into our [backstage](https://backstage.io/) instance (some years ago, we evaluated some enterprise coverage solutions and weren't really satisfied with any of them). We also have a tool that collates coverage with the PR diff to let developers know what percentage of their newly added lines of code are covered (forked from the [diff-cover](https://pypi.org/project/diff-cover) project).

**Reply by sberrevoets** [reply]
Which data do you show on the web UI? The standard HTML output from llvm-cov/JaCoCo or are you parsing the raw coverage files?

**Comment by fdiaz** [comment]
At Airbnb (iOS) we let developers set a minimum required per-project (module) coverage % on their code. They need to specify a number when they're creating a new module but there's no requirement on what that number should be (0% is allowed).

We block a PR if a module goes below the number set. If the user sets a number that's higher that the current coverage level, then a PR shouldn't drop below that number, but it can go up (even if it doesn't reach the defined number).

We also group modules into "Test Levels" depending on their coverage number and tie those with descriptions (e.g. Test Level 1 is for not production ready code).

We also provide coverage reports per PR with a web UI and post-merge teams can look at dashboards to see how their team overall is doing in terms of Test Levels / code coverage where they can dig deeper into each of their projects.

The whole concept of Test Levels addresses something Keith mentions:

> We don't really use coverage as a top level metric for much, just another indicator for quality, but it's mostly a developer facing tool to understand areas that aren't tested but potentially could be.

The existence of Test Levels IMO has helped moved the discussion about code coverage from a purely developer focused discussion to a quality discussion that can engage other areas of the company. There's no full top-down requirement for Test Levels but some areas of the org have been implementing their own local requirements.

**Reply by sberrevoets** [reply]
Yes this is what our workflow is like for the most part as well. Do you have a custom built web UI for this, or is it something like codecov? One of the challenges we're seeing is capturing baseline coverage, so you can see per-PR diffs. Codecov handles this, but doing something in-house means either building out a database with that information or running coverage on both the base branch (or simply the main branch) and the PR'ed branch.

**Reply by fdiaz** [reply]
Yeah, we have an internal tool that handles all of this that is shared across platforms. I don't know a ton about that tool unfortunately but I can always ask!

**Comment by batuhansk** [comment]
At Trendyol (iOS), we use [slather](https://github.com/SlatherOrg/slather) to generate test coverage reports. 

- We run the tests via `xcodebuild` by specifying test plan and scheme, if succeed, then we generate the test coverage reports with [slather](https://github.com/SlatherOrg/slather).
- We post the test coverage result into the private channel on the Slack by using [Slack Webhooks](https://api.slack.com/messaging/webhooks) with several informations. For instance; PR name, test coverage report URL, previous/current coverage percentage, and etc.

At the end, we block the PR from being merged, if coverage percentage drops.

If you would like to learn more about integrating this tool. My colleague ([Beyza](https://github.com/beyzaince)) wrote an article named [`Integrate Slather to Trendyol Project`](https://medium.com/trendyol-tech/integrate-slather-to-trendyol-project-39191cee7b7c). It is definitely worth a look! 👋🏼
